### ISSUE TITLE
Revisiting the Tracker

### ISSUE THREAD

**Issue by Mako88** [dxx-rebirth/dxx-rebirth#529]
Hey @vLKp!

I've recently started playing some Rebirth again, and I'd like to rewrite the tracker. I primarily work with C# these days, so I'm thinking I could build it with .NET Core so it's cross-platform and more easily maintainable than its current state using PHP.

In addition, there are a number of improvements that can be made over the current tracker, starting with the regex parsing of strings it currently does. When the original tracker was made, you raised concerns about using the a=, b=, etc. format for the information Rebirth sends, and I wanted to see what your preferred formatting would be.

**Comment by vLKp** [comment]
You're welcome to write in whatever language(s) you find most comfortable, but I would not recommend C#.  Although the Mono project can run .NET code, I think that .NET's heritage from Microsoft will linger over it for a long time to come.  As a more practical matter, I do not work with C#, and do not maintain an environment that can work with it, so I would not maintain a C#-based tracker.  Every time I have tried to work with Mono has been an exercise in frustration.

As for formatting, I would suggest some form of extensible structured data that allows the tracker to ignore fields it does not understand, so that future upgrades to Rebirth do not unnecessarily break use of old trackers, nor constrain Rebirth not to offer new features in order to avoid breaking the tracker.  I have been considering for some time adding a dependency on a library that could provide a standardized parser, most likely either JSON or YAML.  That could be reused to send the data to the tracker as structured text.

**Comment by Mako88** [comment]
I think JSON is a great idea for tracker communication! It'd make it very easy for the tracker to ignore what it doesn't recognize without breaking anything. I've just finished up getting feature-parity with the PHP tracker and I'm hoping to do some testing this weekend before pushing it live. If you're OK with using JSON as the communication format, I'd love to get the ball rolling on transitioning over to it.

It's unfortunate that you don't work in C#, it's really become my favorite language. I'm happy to continue to maintain it though, and writing it in a language I enjoy will make future me a lot more likely to keep making regular updates lol

**Comment by vLKp** [comment]
If the Linux tooling for C# were decent, I would look at it more seriously.  As is, it's not usable in my opinion.

I'd prefer YAML, but the library I had in mind can emit JSON as a side effect.

**Comment by Mako88** [comment]
The new .NET tracker is now up and running at tracker.dxx-rebirth.com however it looks like there's an issue with hole punching. Unfortunately, I don't have a dev environment currently setup for Rebirth, so I'm not able to debug it very well. From what I can tell, a Rebirth client successfully sends a hole punch request to the tracker, and the tracker successfully sends the client's IP/Port to the game host, but the game host doesn't seem to be sending the hole punch packet to the client.

I suspect this may be an issue with the string encoding or format of the IP/Port from the tracker. If you get time, would you be willing to look into what's going on?

**Comment by vLKp** [comment]
I can try to look at it at some point, but I don't think I ever used a tracker based on this iteration of the protocol, even back when the tracker was in PHP.

**Comment by vLKp** [comment]
I spent some time looking at this today, but the mono project doesn't seem to supply the required compiler, so I can't even try to build the tracker to begin debugging it.  Installing mono dumped 93 (!) programs in `/usr/bin`, but `dotnet` is not one of them.  The only files with `dotnet` in the name at all are some `pkg-config` files.  (I had to disable some gdiplus related functionality just to get it to try to install.  I'm doubtful this is relevant to the lack of a compiler, though.)

How do you build the tracker on Linux?

**Comment by Mako88** [comment]
You can install dotnet with the instructions here: https://docs.microsoft.com/en-us/dotnet/core/install/linux then compile with `dotnet build RebirthTracker.sln`.

I'd recommend installing Visual Studio Code with the C# extension for debugging: https://docs.microsoft.com/en-us/dotnet/core/tutorials/with-visual-studio-code

**Comment by vLKp** [comment]
I knew I needed the command `dotnet` from reading your [build-and-run.sh](https://github.com/Mako88/dxx-tracker/blob/master/build-and-run.sh) script.  I had looked at the install page you referenced, but it has no package instructions for Gentoo, which is why I installed Gentoo's `dev-lang/mono` package, expecting it would contain the compiler tools.  It is described as `Mono runtime and class libraries, a C# compiler/interpreter`.  The manual install instructions from Microsoft all assume use of a prebuilt binary downloaded from Microsoft.

> I'd recommend installing Visual Studio Code with the C# extension for debugging: https://docs.microsoft.com/en-us/dotnet/core/tutorials/with-visual-studio-code

There does not appear to be a Gentoo package for this, either.  For the stated problem, I had planned to use basic print debugging rather than stepping through the code with a debugger, so the lack of a package for this is fine.

**Comment by Mako88** [comment]
I think mono is an older compiler for .NET Framework, whereas .NET Core is a newer package that has much better cross-platform support. I don't know much about Gentoo, but maybe this will help: https://gpo.zugaina.org/dev-dotnet/dotnetcore-sdk-bin/ChangeLog ?

**Comment by vLKp** [comment]
Those appear to be precompiled binaries that are only available for amd64.  I have not found a suitable package that provides the `dotnet` build tool, and thus have been unable to build the tracker to attempt to debug it.

**Comment by Mako88** [comment]
Would you be able to verify Rebirth's behavior with the tracker running at tracker.dxx-rebirth.com?

**Comment by vLKp** [comment]
No.  Attempts to communicate with the public tracker fail, which is why I wanted to run and debug a local tracker.

**Comment by Mako88** [comment]
Howdy @vLKp - I'm once again thinking about fiddling with the tracker, and this time I'd like to make it easily maintainable for you (and anyone else who may come along in the future). I'm considering a React frontend and a Node.js backend all in Typescript. Is that stack something you're familiar with?

**Comment by vLKp** [comment]
I have never worked with any of those.  My browser experience is limited to HTML and straight JavaScript, without any of the fancy third-party libraries (jQuery, AngularJS, etc.).  On the browser side, anything from the last ~8 years is newer than I have had reason to read, much less write.

On the server side, I see that NodeJS is packaged for my distribution, but I've never needed to install it even to run other people's projects, much less used it for anything of mine.  It's very rare that I run any sort of publicly visible server.  For a CGI, I would likely choose Python.  For a daemon, I would grudgingly pick Java.  I use Java rarely and doubt I could even pass an entry-level interview for a Java developer position, but I can modify existing Java programs well enough.

**Comment by Mako88** [comment]
Well I decided to go ahead with the React+Node.js route, just cause I think it's a common enough stack these days that it will probably have the longest lifespan and the broadest reach for potential contributors.

Having said that, getting it running locally is now significantly easier, and I've added instructions to the README - if you ever need to debug something with it and run into any issues, definitely let me know!

I realized we never figured out the issue with hole-punching in the previous tracker iteration. I'll see if I can find someone to help me test (since that's not really something I can test locally) and if it's still an issue with the new rewrite, I'll try to get a Rebirth dev environment setup to see if I can figure out what's going on with it.
